lib/fileset/internal.nix

```
@@ -861,28 +863,55 @@ rec {
   # Type: String -> String -> Path -> Attrs -> FileSet
   _fromFetchGit = function: argument: path: extraFetchGitAttrs:
     let
-      # This imports the files unnecessarily, which currently can't be avoided
-      # because `builtins.fetchGit` is the only function exposing which files are tracked by Git.
-      # With the [lazy trees PR](https://github.com/NixOS/nix/pull/6530),
-      # the unnecessarily import could be avoided.
-      # However a simpler alternative still would be [a builtins.gitLsFiles](https://github.com/NixOS/nix/issues/2944).
-      fetchResult = fetchGit ({
-        url = path;
-      } // extraFetchGitAttrs);
+      # The code path for when isStorePath is true
+      tryStorePath =
+        if pathExists (path + "/.git") then
+          # If there is a `.git` directory in the path,
+          # it means that the path was imported unfiltered into the Nix store.
+          # This function should throw in such a case, because
+          # - `fetchGit` doesn't generally work with `.git` directories in store paths
+          # - Importing the entire path could include Git-tracked files
+          throw ''
+            lib.fileset.${function}: The ${argument} (${toString path}) is a store path within a working tree of a Git repository.
+                This indicates that a source directory was imported into the store using a method such as `import "''${./.}"` or `path:.`.
+                This function currently does not support such a use case, since it currently relies on `builtins.fetchGit`.
+                See https://github.com/NixOS/nix/issues/9292.''
```

Even then, it'd be inefficient to copy at least a whole `.git` into the store every time. Those can be quite big.

```suggestion
                You could make this work by using a fetcher such as `fetchGit` instead of copying the whole repository.
                If you can't avoid copying the repo to the store, see https://github.com/NixOS/nix/issues/9292.''
```

It would be great to link a more useful resource that explains what to do instead, because I don't think anyone should rely on that issue to be fixed, because copying a repo into the store is just not good for performance.

Perhaps sometimes even security, if it actually happened to be local and it has unencrypted un-added secrets, although that's a vaguely recognizable condition, that does not originate from `fileset`, and isn't fixable by `fileset`.

---

Yeah, for now I think your suggestion is good. This is such a rare case that I don't think it's worthwhile writing more about it for the time being.